src/abb.c

```diff
@@ -2,53 +2,631 @@
+	arbol->comparador = comparador;
+	arbol->tamanio = 0;
+
+	return arbol;
+}
+
+abb_t *nodo_insertar_en_lugar_correspondiente(nodo_abb_t *nodoBusqueda,
+					      int (*comparador)(void *, void *),
+					      nodo_abb_t *nodoAinsertar,
+					      abb_t *arbol)
+{
+	if (nodoBusqueda == NULL || comparador == NULL) {
+		return NULL;
+	}
+
+	int comp = comparador(nodoAinsertar->elemento, nodoBusqueda->elemento);
+
+	if (comp <= 0) {
+		if (nodoBusqueda->izquierda == NULL) {
+			nodoBusqueda->izquierda = nodoAinsertar;
+			arbol->tamanio += 1;
+			return arbol;
+		}
+
+		return nodo_insertar_en_lugar_correspondiente(
+			nodoBusqueda->izquierda, arbol->comparador,
+			nodoAinsertar, arbol);
+
+	} else if (comp > 0) {
+		if (nodoBusqueda->derecha == NULL) {
+			nodoBusqueda->derecha = nodoAinsertar;
+			arbol->tamanio += 1;
+			return arbol;
+		}
+		return nodo_insertar_en_lugar_correspondiente(
+			nodoBusqueda->derecha, arbol->comparador, nodoAinsertar,
+			arbol);
+	}
 	return NULL;
 }
 
 abb_t *abb_insertar(abb_t *arbol, void *elemento)
 {
+	if (arbol == NULL || arbol->comparador == NULL) {
+		return NULL;
+	}
+
+	nodo_abb_t *nodo = nodo_abb_crear(elemento);
+	if (nodo == NULL) {
+		return NULL;
+	}
+
+	//cAMBIOO
+	if (arbol->nodo_raiz == NULL) {
+		arbol->nodo_raiz = nodo;
+		arbol->tamanio += 1;
+		return arbol;
+
+	} else {
+		return nodo_insertar_en_lugar_correspondiente(
+			arbol->nodo_raiz, arbol->comparador, nodo, arbol);
+	}
 	return arbol;
 }
 
+//Busca el nodo anterior al nodo ingresado dentro del arbol ingresado.
+//devuelve NULL si no existe el nodo
+
+nodo_abb_t *buscarNodoPadre(abb_t *arbol, nodo_abb_t *nodoAbuscar,
+			    nodo_abb_t *nodoActual)
+{
+	if (arbol == NULL || nodoAbuscar == NULL || nodoActual == NULL) {
+		return NULL;
+	}
+
+	if (nodoAbuscar == arbol->nodo_raiz) {
+		return arbol->nodo_raiz;
+	}
+
+	int comp =
+		arbol->comparador(nodoAbuscar->elemento, nodoActual->elemento);
+
+	if (comp <= 0) {
+		if (nodoActual->izquierda == nodoAbuscar) {
+			return nodoActual;
+		}
+		return buscarNodoPadre(arbol, nodoAbuscar,
+				       nodoActual->izquierda);
+
+	} else if (comp > 0) {
+		if (nodoActual->derecha == nodoAbuscar) {
+			return nodoActual;
+		}
+		return buscarNodoPadre(arbol, nodoAbuscar, nodoActual->derecha);
+	}
+	return NULL;
+}
+
+nodo_abb_t *nodo_reemplazo(nodo_abb_t *nodoActual)
+{
+	if (nodoActual == NULL) {
+		return NULL;
+	}
+
+	if (nodoActual->derecha == NULL) {
+		return nodoActual;
+	}
+
+	return nodo_reemplazo(nodoActual->derecha);
+}
+
+//Elimina un nod que tnega dos hijos, contemplando el caso de que sea raiz.
+void abb_quitar_con_2_hijos(nodo_abb_t *nodoAEliminar, abb_t *arbol)
+{
+	if (nodoAEliminar == NULL || arbol == NULL) {
+		return;
+	}
+
+	//Va una posicion hacia la izquierda y luego dentro de la funcion,
+	//se desplaza hacia la derecha hasta encontarar un nodo que no tenga
+	//un hijo hacia la derecha.
+	nodo_abb_t *nodoReemplazo = nodo_reemplazo(nodoAEliminar->izquierda);
```

Comment:
Buscar el predecesor inorden se puede modularizar para que no te quede una función tan compleja y larga.

Reply:
Me corrijo: El predecesor in orden lo buscás fácilmente pero después te complicas mucho con el resto de la lógica y se hace difícil entenderlo fácilmente.